File: src/main/java/io/jenkins/plugins/gamekins/property/GameOrganizationFolderProperty.java
package io.jenkins.plugins.gamekins.property;

import com.cloudbees.hudson.plugins.folder.AbstractFolder;
import com.cloudbees.hudson.plugins.folder.AbstractFolderProperty;
import com.cloudbees.hudson.plugins.folder.AbstractFolderPropertyDescriptor;
import hudson.Extension;
import hudson.model.AbstractItem;
import hudson.util.ListBoxModel;
import jenkins.branch.MultiBranchProject;
import jenkins.branch.OrganizationFolder;
import net.sf.json.JSONObject;
import org.kohsuke.stapler.*;

import javax.annotation.Nonnull;
import java.io.IOException;

public class GameOrganizationFolderProperty extends AbstractFolderProperty<AbstractFolder<?>> {

    private GameOrganizationFolderProperty() { }

    @Extension
    public static class DescriptorFolderImpl extends AbstractFolderPropertyDescriptor {

        public DescriptorFolderImpl() {
            super();
            load();
        }

        @Nonnull
        @Override
        public String getDisplayName() {
            return "Set the activation of the Gamekins plugin.";
        }

        @Override
        public boolean isApplicable(Class<? extends AbstractFolder> containerType) {
            return containerType == OrganizationFolder.class;
        }

        @Override
        public AbstractFolderProperty<?> newInstance(StaplerRequest req, JSONObject formData) {
            if (req == null || formData == null) return null;
            OrganizationFolder folder = (OrganizationFolder) req.findAncestor(OrganizationFolder.class).getObject();
            for (MultiBranchProject<?, ?> project: folder.getItems()) {
                if (project.getName().equals(formData.getString("project"))) {
                    try {
                        GameMultiBranchProperty property = project.getProperties().get(GameMultiBranchProperty.class);
                        if (property == null) {
                            project.addProperty(new GameMultiBranchProperty(project,
                                    formData.getBoolean("activated"), formData.getBoolean("showStatistics")));
                        } else {
                            property.reconfigure(req, formData);
                        }
                        folder.save();
                        break;
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
            }
            return null;
        }

        public ListBoxModel doFillProjectItems(@AncestorInPath OrganizationFolder job) {
            if (job == null) return new ListBoxModel();
            ListBoxModel listBoxModel = new ListBoxModel();
            job.getItems().stream().map(AbstractItem::getName).forEach(listBoxModel::add);
            return listBoxModel;
        }
    }
}

File: src/main/kotlin/io/jenkins/plugins/gamekins/property/GameOrganizationFolderProperty.kt
package io.jenkins.plugins.gamekins.property

import com.cloudbees.hudson.plugins.folder.AbstractFolder
import com.cloudbees.hudson.plugins.folder.AbstractFolderProperty
import com.cloudbees.hudson.plugins.folder.AbstractFolderPropertyDescriptor
import hudson.Extension
import hudson.util.ListBoxModel
import jenkins.branch.MultiBranchProject
import jenkins.branch.OrganizationFolder
import net.sf.json.JSONObject
import org.kohsuke.stapler.AncestorInPath
import org.kohsuke.stapler.StaplerRequest
import java.io.IOException
import javax.annotation.Nonnull

class GameOrganizationFolderProperty private constructor() : AbstractFolderProperty<AbstractFolder<*>?>() {

    @Extension
    class DescriptorFolderImpl : AbstractFolderPropertyDescriptor() {
        @Nonnull
        override fun getDisplayName(): String {
            return "Set the activation of the Gamekins plugin."
        }

        override fun isApplicable(containerType: Class<out AbstractFolder<*>?>): Boolean {
            return containerType == OrganizationFolder::class.java
        }

        override fun newInstance(req: StaplerRequest?, formData: JSONObject): AbstractFolderProperty<*>? {
            if (req == null) return null
            val folder = req.findAncestor(OrganizationFolder::class.java).getObject() as OrganizationFolder
            for (project in folder.items) {
                if (project.name == formData.getString("project")) {
                    try {
                        val property = project.properties.get(GameMultiBranchProperty::class.java)
                        property?.reconfigure(req, formData)
                                ?: project.addProperty(GameMultiBranchProperty(project,
                                        formData.getBoolean("activated"), formData.getBoolean("showStatistics")))
                        folder.save()
                        break
                    } catch (e: IOException) {
                        e.printStackTrace()
                    }
                }
            }
            return null
        }

        fun doFillProjectItems(@AncestorInPath job: OrganizationFolder?): ListBoxModel {
            if (job == null) return ListBoxModel()
            val listBoxModel = ListBoxModel()
            job.items.stream().map { obj: MultiBranchProject<*, *> -> obj.name }
                    .forEach { nameAndValue: String? -> listBoxModel.add(nameAndValue) }
            return listBoxModel
        }

        init {
            load()
        }
    }
}
